File: strands-py/src/strands/vended_tools/a2a_client/a2a_client.py
"""A2A client tool for communicating with remote A2A-protocol agents.

Provides :func:`make_a2a_client`, a factory that requires an explicit mapping of
permitted endpoints to their :class:`~a2a.client.ClientConfig`, plus optional size limits.

The tool is a stateless shim over :class:`~strands.agent.a2a_agent.A2AAgent`.
A fresh ``A2AAgent`` is constructed on every call so the tool carries no session
state between invocations.  Each endpoint may carry its own
:class:`~a2a.client.ClientConfig` to support per-endpoint authentication
(bearer tokens, SigV4, OAuth).
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any, Literal

try:
    from a2a.client import ClientConfig

    from ...agent.a2a_agent import A2AAgent
except ImportError as error:
    raise ImportError("a2a_client requires the 'a2a' extra. Install with: pip install 'strands-agents[a2a]'") from error
from ...tools.decorator import tool

if TYPE_CHECKING:
    from ...tools.decorator import DecoratedFunctionTool

_DEFAULT_MAX_BYTES = 5 * 1024 * 1024

_A2AClientOutput = dict[str, Any]

DEFAULT_A2A_CLIENT_DESCRIPTION = (
    "Interacts with remote A2A (Agent-to-Agent) protocol agents. "
    "Use operation='discover' to fetch an agent card from an endpoint. "
    "Use operation='send_message' to send a message and receive a response. "
    "Only the listed endpoints are permitted."
)


class A2AClientError(RuntimeError):
    """Raised when an A2A operation fails."""


def make_a2a_client(
    *,
    name: str = "a2a_client",
    description: str | None = None,
    allowed_endpoints: dict[str, ClientConfig | None],
    max_bytes: int = _DEFAULT_MAX_BYTES,
) -> DecoratedFunctionTool:
    """Create an A2A client tool.

    Args:
        name: Tool name shown to the model.
        description: Tool description shown to the model. When ``None``,
            generated from ``DEFAULT_A2A_CLIENT_DESCRIPTION`` plus the
            permitted endpoints list.
        allowed_endpoints: Mapping of permitted base URLs to their
            :class:`~a2a.client.ClientConfig`.  Use ``None`` as the value for
            endpoints that need no custom configuration. Any endpoint not in this
            mapping is rejected before a network connection is made.
        max_bytes: Maximum size in bytes of the result dict returned to the model.
            Does not cap the network transfer or binary parts.
            Results larger than this cap are rejected with an error.

    Returns:
        A decorated tool that communicates with A2A agents.
    """
    if not allowed_endpoints:
        raise ValueError("allowed_endpoints must contain at least one endpoint")
    if max_bytes <= 0:
        raise ValueError(f"max_bytes must be positive, got {max_bytes}")

    if description is None:
        endpoints_list = ", ".join(sorted(allowed_endpoints))
        description = f"{DEFAULT_A2A_CLIENT_DESCRIPTION} Permitted endpoints: {endpoints_list}."

    @tool(name=name, description=description)
    async def a2a_client_tool(
        operation: Literal["discover", "send_message"],
        endpoint: str,
        message: str | None = None,
    ) -> _A2AClientOutput:
        """Interact with a remote A2A-protocol agent.

        Args:
            operation: Action to perform — ``discover`` to fetch the agent card,
                or ``send_message`` to send a text message and receive a response.
            endpoint: Base URL of the target A2A agent.
            message: Text to send to the agent.  Required when
                ``operation`` is ``send_message``; ignored otherwise.

        Returns:
            Result dict from the A2A operation.

        Raises:
            A2AClientError: When the endpoint is not in the allowlist, when
                ``message`` is missing for ``send_message``, or when the
                underlying A2A call fails.
        """
        # Check if the endpoint is allowed via exact-match.
        if endpoint not in allowed_endpoints:
            raise A2AClientError(
                f"Endpoint '{endpoint}' is not in the allowed endpoints list. "
                f"Permitted endpoints: {sorted(allowed_endpoints)}"
            )

        agent = A2AAgent(endpoint, client_config=allowed_endpoints[endpoint])

        if operation == "discover":
            return await _handle_discover(agent, max_bytes)

        if operation == "send_message":
            if not message:
                raise A2AClientError("'message' is required for send_message operation")
            return await _handle_send_message(agent, message, max_bytes)

        raise A2AClientError(f"Unknown operation: {operation!r}")

    return a2a_client_tool


async def _handle_discover(agent: A2AAgent, max_bytes: int) -> _A2AClientOutput:
    """Fetch the agent card via *agent* and return it as a dict."""
    try:
        agent_card = await agent.get_agent_card()
    except Exception as error:
        raise A2AClientError(f"Failed to discover agent card at {agent.endpoint!r}: {error}") from error

    result: dict[str, Any] = agent_card.model_dump(mode="json", exclude_none=True)
    size = len(json.dumps(result).encode())
    if size > max_bytes:
        raise A2AClientError(f"Agent card response exceeds max_bytes limit ({size} > {max_bytes})")
    return result


async def _handle_send_message(agent: A2AAgent, message_text: str, max_bytes: int) -> _A2AClientOutput:
    """Send *message_text* via *agent* and return the response as a dict."""
    try:
        agent_result = await agent.invoke_async(message_text)
    except Exception as error:
        raise A2AClientError(f"Failed to send message to {agent.endpoint!r}: {error}") from error

    task_state: str | None = agent_result.state.get("a2a_task_state") if agent_result.state else None
    if task_state and task_state != "completed":
        detail = " ".join(block["text"] for block in agent_result.message["content"] if "text" in block)
        raise A2AClientError(
            f"Remote agent at {agent.endpoint!r} did not complete: task state is {task_state!r}. {detail}".rstrip()
        )

    result: dict[str, Any] = {"message": agent_result.message}
    size = len(json.dumps(result).encode())
    if size > max_bytes:
        raise A2AClientError(f"Response exceeds max_bytes limit ({size} > {max_bytes})")
    return result

File: strands-ts/src/vended-tools/a2a-client/a2a-client.ts
import type { ClientFactory as ClientFactoryType } from '@a2a-js/sdk/client'
import type { AgentCard, TextPart } from '@a2a-js/sdk'
import { tool } from '../../tools/tool-factory.js'
import type { JSONValue } from '../../types/json.js'
import type { MessageData } from '../../types/messages.js'
import { A2AAgent } from '../../a2a/a2a-agent.js'
import { A2AStreamUpdateEvent } from '../../a2a/events.js'
import { z } from 'zod'

const DEFAULT_MAX_BYTES = 5 * 1024 * 1024

/**
 * Zod schema for A2A client input validation.
 */
const a2aClientInputSchema = z.object({
  operation: z
    .enum(['discover', 'send_message'])
    .describe("Action to perform — 'discover' to fetch the agent card, or 'send_message' to send a message"),
  endpoint: z.string().describe('Base URL of the target A2A agent. Must be one of the permitted endpoints.'),
  message: z
    .string()
    .nullable()
    .optional()
    .describe("Text to send to the agent. Required when operation is 'send_message'; ignored otherwise."),
})

/**
 * Default description shown to the model for the A2A client tool.
 *
 * @example
 * ```typescript
 * import { ClientFactory, DefaultAgentCardResolver, JsonRpcTransportFactory, RestTransportFactory, createAuthenticatingFetchWithRetry } from '@a2a-js/sdk/client'
 *
 * const authFetch = createAuthenticatingFetchWithRetry(fetch, {
 *   headers: async () => ({ Authorization: 'Bearer your-token' }),
 *   shouldRetryWithHeaders: async () => undefined,
 * })
 *
 * const a2aClient = makeA2AClient({
 *   allowedEndpoints: {
 *     'https://agent.example.com': undefined,
 *     'https://secure-agent.example.com': new ClientFactory({
 *       transports: [
 *         new JsonRpcTransportFactory({ fetchImpl: authFetch }),
 *         new RestTransportFactory({ fetchImpl: authFetch }),
 *       ],
 *       cardResolver: new DefaultAgentCardResolver({ fetchImpl: authFetch }),
 *     }),
 *   },
 * })
 * const agent = new Agent({ model, tools: [a2aClient] })
 * ```
 */
export const DEFAULT_A2A_CLIENT_DESCRIPTION =
  'Interacts with remote A2A (Agent-to-Agent) protocol agents. ' +
  "Use operation='discover' to fetch an agent card from an endpoint. " +
  "Use operation='send_message' to send a message and receive a response. " +
  'Only the listed endpoints are permitted.'

export interface MakeA2AClientOptions {
  name?: string
  description?: string
  allowedEndpoints: Record<string, ClientFactoryType | undefined>
  maxBytes?: number
}

/**
 * Create an A2A client tool that communicates with remote A2A-protocol agents.
 * Each endpoint may carry its own ClientFactory for per-endpoint authentication.
 * A fresh A2AAgent is constructed on every call (stateless).
 */
export function makeA2AClient(options: MakeA2AClientOptions): ReturnType<typeof tool> {
  const endpoints = Object.keys(options.allowedEndpoints)
  if (endpoints.length === 0) {
    throw new Error('allowedEndpoints must contain at least one endpoint')
  }

  const maxBytes = options.maxBytes ?? DEFAULT_MAX_BYTES
  if (!Number.isInteger(maxBytes) || maxBytes <= 0) {
    throw new Error(`maxBytes must be a positive integer, got ${String(maxBytes)}`)
  }

  const description =
    options.description ?? `${DEFAULT_A2A_CLIENT_DESCRIPTION} Permitted endpoints: ${[...endpoints].sort().join(', ')}.`

  return tool({
    name: options.name ?? 'a2a_client',
    description,
    inputSchema: a2aClientInputSchema,
    callback: async (input) => {
      const { operation, endpoint, message } = input

      if (!Object.hasOwn(options.allowedEndpoints, endpoint)) {
        throw new Error(
          `Endpoint '${endpoint}' is not in the allowed endpoints list. ` +
            `Permitted endpoints: ${[...endpoints].sort().join(', ')}`
        )
      }

      const clientFactory = options.allowedEndpoints[endpoint]
      const agent = new A2AAgent({
        url: endpoint,
        ...(clientFactory !== undefined ? { clientFactory } : {}),
      })

      if (operation === 'discover') {
        return handleDiscover(agent, maxBytes) as unknown as Promise<JSONValue>
      }

      if (operation === 'send_message') {
        if (!message) {
          throw new Error("'message' is required for send_message operation")
        }
        return handleSendMessage(agent, message, maxBytes) as unknown as Promise<JSONValue>
      }

      throw new Error(`Unknown operation: '${String(operation)}'`)
    },
  })
}

async function handleDiscover(agent: A2AAgent, maxBytes: number): Promise<AgentCard> {
  let agentCard: AgentCard
  try {
    agentCard = await agent.getAgentCard()
  } catch (error) {
    throw new Error(`Failed to discover agent card at '${agent.id}': ${String(error)}`, { cause: error })
  }

  const size = new TextEncoder().encode(JSON.stringify(agentCard)).length
  if (size > maxBytes) {
    throw new Error(`Agent card response exceeds maxBytes limit (${size} > ${maxBytes})`)
  }
  return agentCard
}

async function handleSendMessage(
  agent: A2AAgent,
  message: string,
  maxBytes: number
): Promise<{ message: MessageData }> {
  let taskState: string | undefined
  let statusText = ''
  let resultMessage: MessageData

  try {
    const gen = agent.stream(message)
    let next = await gen.next()
    while (!next.done) {
      const event = next.value instanceof A2AStreamUpdateEvent ? next.value.event : undefined
      if (event?.kind === 'task' || event?.kind === 'status-update') {
        taskState = event.status.state
        statusText = (event.status.message?.parts ?? [])
          .filter((p): p is TextPart => p.kind === 'text')
          .map((p) => p.text)
          .join(' ')
      }
      next = await gen.next()
    }
    const { role, content } = next.value.lastMessage.toJSON()
    resultMessage = { role, content }
  } catch (error) {
    throw new Error(`Failed to send message to '${agent.id}': ${String(error)}`, { cause: error })
  }

  if (taskState !== undefined && taskState !== 'completed') {
    throw new Error(
      `Remote agent at '${agent.id}' did not complete: task state is '${taskState}'. ${statusText}`.trimEnd()
    )
  }

  const result = { message: resultMessage }
  const size = new TextEncoder().encode(JSON.stringify(result)).length
  if (size > maxBytes) {
    throw new Error(`Response exceeds maxBytes limit (${size} > ${maxBytes})`)
  }
  return result
}
